DG Appendix: Manual Testing (#286)
AY2122S1-CS2103T-W08-3/tp: docs/DeveloperGuide.md

### Display client's total orders feature

The feature displays the total orders for all clients except those without orders in a new window. 
Its mechanism is a mix of the mechanisms for `MainWindow` and `HelpWindow`.

Similar to `help` and `exit`, `CommandResult` has a dedicated `boolean` field to indicate whether the command is a 
`totalorders` command. There is also a dedicated method to handle `totalorders` command in `MainWindow` class.
By calling this method, the data of total orders is reloaded similar to loading other data (`Client/Task/Order`) in the Main Window, i.e. 
through the `Logic` component. After reloading data, that method shows or focuses the total orders window similar to that of the 
help window.

#### Execution

The sequence diagram below shows the interaction within the `UI` component when a `totalorders` command is executed.

> ![Interactions Inside the Logic and Model Component for the `totalorders` Command](images/TotalOrdersSequenceDiagram2.png)

Reviewer: ```suggestion
![Interactions within the UI Component for the `totalorders` Command](images/TotalOrdersSequenceDiagram2.png)
```

Author: Thanks for pointing out the incorrect component, but I think using _Inside_ is fine.

The sequence diagram below shows the interaction within the `Logic` component when the `UI` component calls 
`execute("totalorders")`. Note that there is no need to have a `TotalOrdersCommandParser`. This is because the 
`SalesNoteParser` can directly create and return a `TotalOrdersCommand`, similar to that of `help` and `exit` commands.  

> ![Interactions Inside the Logic and Model Component for the `totalorders` Command](images/TotalOrdersSequenceDiagram1.png)

Reviewer: ```suggestion
![Interactions within the Logic Component for the `totalorders` Command](images/TotalOrdersSequenceDiagram1.png)
```

Author: Thanks for pointing out the incorrect component, but I think using _Inside_ is fine.

   1. Prerequisites: List all clients using the `listclients` command. Multiple clients in the list.

   1. Test case: `deleteclient 1`<br>
      Expected: First client is deleted from the list. Details of the deleted client shown in the status message.

   1. Test case: `deleteclient 0`<br>
      Expected: No client is deleted. Error details shown in the status message. Status bar remains the same.

   1. Other incorrect delete client commands to try: `deleteclient`, `deleteclient x`, `...` (where x is larger than the list size)<br>
      Expected: Similar to previous.

1. Deleting a client while no clients are being shown

   1. Prerequisite: Use `findclients keyword` command (where keyword matches none of the clients) to display an empty list of clients.<br>

   1. Test case: `deleteclient 1`<br>
      Expected: No client is deleted. Error details shown in the status message. Status bar remains the same.

   1. Other delete client commands to try: `deleteclient`, `deleteclient x`, `...` (where x is anything)<br>
      Expected: Similar to previous. Note that there is no valid command because the client list is empty.

### Saving data

1. Dealing with missing data files

>    1. Prerequisite: The folder containing the jar file contains a `data` folder.

Reviewer: The prerequisite can be more explicit, e.g., '.... contains an empty `data` folder.'

Author: This test case deals with missing data. Regardless whether the `data` folder is empty, we can delete the `data` folder to simulate missing data (refer to the next line). Hence, this description is intentional, since it does not matter whether the `data` folder is empty.

   1. Delete the `data` folder then launch the app by double-clicking the jar file.<br>
      Expected: Shows the GUI with a set of clients, tasks and orders. The window size may not be optimum.

1. Dealing with corrupted data files

>    1. Prerequisite: The folder containing the jar file contains a `data` folder, which contains `addressbook.json`, `salesBook.json` and `taskBook.json`. 

Reviewer: The JSON file for Addressbook should be `addressBook.json` instead of `addressbook.json`.

Author: Thanks for pointing out!

>    1. Test case: Corrupt the `addressbook.json` file by deleting a closing curly bracket (`}`) or any other brackets.<br>

Reviewer: `addressBook.json`

Author: Thanks for pointing out!

>       Expected: Shows the GUI with no clients, but with a previously saved set of tasks and orders.

Reviewer: This is not true because there is a possibility that the orders/tasks are wiped too if the customer of an order does not exist in the new 'empty' client list. 
If I remember correctly, there are two possibilities for this case (all data is wiped (if any order exists) or only previous tasks persist (if all the tasks are tagged with `General`))

Author: Hmm seems like the behavior changed. I will update it. Thanks for pointing out.

Author: I tried, even if all tasks are tagged with `General`, they will also be wiped out.

Reviewer: Correction:
If I remember correctly, there are two possibilities for this case (all data is wiped (if any order exists) or only previous tasks persist (if all the tasks are tagged with General and the order list is empty))

Reviewer: This is actually an implementation bug. It should behave like what I stated.

Reviewer: You may want to wait until #294 to be merged so that you can observe the exact behaviour.

Author: Okay, I will update it now.

      Expected: Similar to previous. Note that there is no valid command because the client list is empty.

### Saving data

1. Dealing with missing data files

   1. Prerequisite: The folder containing the jar file contains a `data` folder.

   1. Delete the `data` folder then launch the app by double-clicking the jar file.<br>
      Expected: Shows the GUI with sample clients, tasks and orders. The window size may not be optimum.

1. Dealing with corrupted data files

   1. Prerequisite: The folder containing the jar file contains a `data` folder, which contains `addressBook.json`, `taskBook.json` and `salesBook.json`. 

   1. Test case: Corrupt the `addressBook.json` file by deleting a closing curly bracket (`}`) or any other brackets.<br>
      Expected: Shows the GUI without any data.
      
   1. Test case: Corrupt the `taskBook.json` file by deleting a closing curly bracket (`}`) or any other brackets.<br>
      Expected: Shows the GUI without any tasks, but with previously saved clients and orders.

   1. Test case: Corrupt the `salesBook.json` file by deleting a closing curly bracket (`}`) or any other brackets.<br>

>       Expected: Shows the GUI without any data.

Reviewer: In case you hadn't noticed, this is not the only expected result too.

   1. Prerequisite: Use `findclients keyword` command (where keyword matches none of the clients) to display an empty list of clients.<br>

   1. Test case: `deleteclient 1`<br>
      Expected: No client is deleted. Error details shown in the status message. Status bar remains the same.

   1. Other delete client commands to try: `deleteclient`, `deleteclient x`, `...` (where x is anything)<br>
      Expected: Similar to previous. Note that there is no valid command because the client list is empty.

### Saving data

1. Dealing with missing data files

   1. Prerequisite: The folder containing the jar file contains a `data` folder.

   1. Delete the `data` folder then launch the app by double-clicking the jar file.<br>
      Expected: Shows the GUI with sample clients, tasks and orders. The window size may not be optimum.

1. Dealing with corrupted data files

   1. Prerequisites: The folder containing the jar file contains a `data` folder, which contains `addressBook.json`,  
      `taskBook.json` and `salesBook.json`.

   1. Test case: Corrupt the `addressBook.json` file by deleting a closing curly bracket (`}`) or any other brackets.<br>

>       Expected: Shows the GUI without any data.

Reviewer: Perhaps you can divide it into 2 scenarios here too.

Author: Done.
